File: CompetitiveProgramming/Arrays/grid_challenge.rb
#!/usr/bin/env ruby

def grid_challenge(grid)
  # First, sort each of the strings we get.
  grid.map! do |str|
    str.chars.sort
  end

  # Then, we could iterate the columns with indices, or transpose the matrix
  # to form "new strings" with the columns rather than the rows. And from there
  # simply check whether they are sorted, and return the "YES" or the "NO" accordingly.

  grid.transpose.each do |column|
    0.upto(column.length - 2) do |i|
      if column[i] >= column[i + 1] then return "NO" end
    end
  end

  "YES"
end

# Available Test Cases:
#
# Input: ["eabcd", "fghij", "olkmn", "trpqs", "xywuv"]
# Output "YES"

File: CompetitiveProgramming/Hackerrank/Easy/grid_challenge.py
#!/usr/env/bin python3

def grid_challenge(grid: list[str]) -> str:
    # First we have to sort the rows.

    for i in range(0, len(grid)):
        grid[i] = sorted(grid[i])

    # And here, we are doing an "inversed" double loop so that we can iterate
    # over the columns, rather than the rows, to check whether they are sorted.

    for i in range(0, len(grid[0])):
        for j in range(0, len(grid) - 1):
            if grid[j][i] > grid[j+1][i]:
                return "NO"

    return "YES"


def main():
    t = int(input().strip())

    for _ in range(t):
        n = int(input().strip())
        grid = []

        for _ in range(n):
            grid_item = input()
            grid.append(grid_item)

        result = grid_challenge(grid)
        print(result)


if __name__ == '__main__':
    main()
